packages/agreement-process/test/agreement.integration.test.ts

```diff
@@ -0,0 +1,713 @@
+
+      const eservice = buildEService(
+        generateId<EServiceId>(),
+        eserviceProducer.id,
+        [descriptor]
+      );
+
+      await writeInReadmodel<Tenant>(eserviceProducer, tenants);
+      await writeInReadmodel<Tenant>(consumer, tenants);
+      await writeInReadmodel<EService>(eservice, eservices);
+
+      const authData = getRandomAuthData(consumer.id); // different from eserviceProducer
+      const apiAgreementPayload: ApiAgreementPayload = {
+        eserviceId: eservice.id,
+        descriptorId: eservice.descriptors[0].id,
+      };
+
+      const createdAgreementId = await agreementService.createAgreement(
+        apiAgreementPayload,
+        authData
+      );
+
+      await expectedAgreementCreation(
+        unsafeBrandId<AgreementId>(createdAgreementId),
+        eservice.id,
+        descriptor.id,
+        eserviceProducer.id,
+        consumer.id
+      );
+    });
+
+    it("should succeed when EService's latest Descriptors are draft, and the latest non-draft Descriptor is published", async () => {
+      const tenant: Tenant = buildTenant();
+
+      const descriptor0: Descriptor = buildDescriptorPublished();
+      const descriptor1: Descriptor = {
+        ...buildDescriptorPublished(),
+        version: "1",
+        state: descriptorState.draft,
+      };
+
+      const descriptor2: Descriptor = {
+        ...buildDescriptorPublished(),
+        version: "2",
+        state: descriptorState.draft,
+      };
+
+      const eservice = buildEService(generateId<EServiceId>(), tenant.id, [
+        descriptor0,
+        descriptor1,
+        descriptor2,
+      ]);
+
+      await writeInReadmodel<Tenant>(tenant, tenants);
+      await writeInReadmodel<EService>(eservice, eservices);
+
+      const authData = getRandomAuthData(tenant.id);
+      const apiAgreementPayload: ApiAgreementPayload = {
+        eserviceId: eservice.id,
+        descriptorId: descriptor0.id,
+      };
+
+      const createdAgreementId = await agreementService.createAgreement(
+        apiAgreementPayload,
+        authData
+      );
+
+      await expectedAgreementCreation(
+        unsafeBrandId<AgreementId>(createdAgreementId),
+        eservice.id,
+        descriptor0.id,
+        tenant.id,
+        tenant.id
+      );
+    });
+
+    it("should succeed when Agreements in non-conflicting states exist for the same EService and consumer", async () => {
+      const tenant: Tenant = buildTenant();
+      const descriptor: Descriptor = buildDescriptorPublished();
+
+      const eservice = buildEService(generateId<EServiceId>(), tenant.id, [
+        descriptor,
+      ]);
+
+      const otherAgreement = buildAgreementWithValidCreationState(
+        eservice.id,
+        tenant.id
+      );
+
+      await writeInReadmodel<Tenant>(tenant, tenants);
+      await writeInReadmodel<EService>(eservice, eservices);
+      await writeInReadmodel<Agreement>(otherAgreement, agreements);
```

Comment:
To improve data coherence, when we add an `agreement` we should also add it into the `event-store` with a `creation` event. Instead, for `tenant` and `eService`, adding into `readModel` is enough in this test suite.

Reply:
If I understood correctly you suggest adding them also in the "event store" as well as in the "read-model" 
**IMHO** 
the additional insert into `event-store` of all event's history, reflects the real status more deeply, anyway storing all previous events of "collateral" entities in event-store shouldn't interest process computation in some way; The process services only writes and never reads from `event-store`, for our design the read-model is the data sources truth and added events was never read form it.
*The only useful case are when we have a previous events on the same entity, for example during agreement upgrade, in this case we can also check that event's versioning is never violated during writing.

Definitively I suggest to not overload the integration test with IO ops with a with dubious relevance for testing.